### ISSUE TITLE
Question on running shellcheck on files in directory like /usr/bin

### ISSUE THREAD

**Issue by msetzerii** [koalaman/shellcheck#2079]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or "online"): 
- [ ] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:
Recently discovered a bug in /usr/bin/firefox script that caused a zombie
In discussion on list, it seemed they bug was known, and a fix came out, 
but the fix had an error with a missing space after a [ 
A user mentioned your program, and I ran it on the script, and it points out some other issues as well.

Was interested in seeing if other scripts in /usr/bin might have issues, so created a little script.
#!/usr/bin/bash
for a in *; do file "$a" | grep "sh script" | cut -f 1 -d: ; done >file-list
#for a in $(<file-list) ; do shellcheck -Serror "$a" ; done >>errors-found
shellcheck -Serror $(<file-list) >errors-found

My /usr/bin has about 4800 files, but only about 400 are actually shell script.
The first part finds all the files that have sh script in the result.
This presents another question. Fedora has /usr/bin/sh actually point to bash, but scripts have shell listed as sh??

Running my script, it produces almost 700 lines of errors from all the scripts.
Run it with forced -s bash, and there were only a few differences in  lines.

Wondering if there is a better way to test these things. Was surprised with how many errors showed up?

```sh

#!/your/interpreter
your script here

Again, just found out of shellcheck, and it is very interesting.
Check some of my own scripts, and no errors found, but a number of warnings, that I took the suggestions to update code, so now they come up clean.


```

#### Here's what shellcheck currently says:
Thanks.

#### Here's what I wanted or expected to see:
Know my little script is very basic, but it does point out some issues that should probable be fixed in all the shell scripts.
Since these all come from Fedora, would have expected them to be as correct as possible.

**Comment by marc-h38** [comment]
> Running my script, it produces almost 700 lines of errors from all the scripts.
> Run it with forced -s bash, and there were only a few differences in lines.

So you're saying shellcheck identification differs from the identification performed by the `file` command, is that the shellcheck issue you're reporting?  If yes can you focus on a couple specific examples? It's not very clear what the bug/question is.

> Since these all come from Fedora, would have expected them to be as correct as possible.

This bug tracker is not about Fedora.

**Comment by msetzerii** [comment]
Know it isn't about Fedora, but is about using shellcheck to look for problems.
It states that it uses different rules based on the shell being defined in the script being checked.
With the example of Fedora: Some scripts list the shell at beginning as being sh, but on Fedora sh is linked to bash, so the rules being used wouldn't truly match, since something that might be wrong for sh, but fine for bash would be compared as sh?
Same would apply for distros that might map sh to dash or other shell.

Suppose one could just use the option to force the use of shell bash rather than the default of using what the script shows.
Was just thinking of doing something to check if sh is actually sh, or if it was linked to some other shell, and thus using the actual one. Guess it isn't a big issue?? Thanks for your time.

**Comment by brother** [comment]
I read the shebang as a lowest common denominator. When it is `sh` it is supposed to be working with anything that is fully POSIX shell - dash, bash and others. If you state `bash` I would not count on that it would work when executed with busybox sh or dash and such.

Read the shebang as a recommendation from the author and a hint about how protable the script is supposed to be rather than what shell interpreter you must use.

If you move a fedora sh tagged script into my debian system it will be executed with dash, and probably be fine with regards to shell syntax - other stuff would probably break.

**Comment by marc-h38** [comment]
> I read the shebang as a lowest common denominator. When it is sh it is supposed to be working with anything that is fully POSIX shell - dash, bash and others. 

Exactly that. If a script requires bash then it should start with `#!/bin/bash` or `#!/usr/bin/env bash`. Starting a script with `#!/bin/sh` and using bashisms in that script is a bug, don't do that. If Fedora does that then it's a bug in Fedora. You already found the perfect workaround: `shellcheck -s `

By the way bash behaves slightly differently when invoked with `#!/bin/sh`, see `info bash`.

> This presents another question.

What was the previous question? I see an anecdote about Firefox but no question. The title of this bug has nothing specific topic either.

**Comment by TinCanTech** [comment]
> The title of this bug has nothing specific topic either.

The title of this bug has nothing topic specific either

**Comment by marc-h38** [comment]
Thanks for correcting me,  actually I meant: The title of this bug has no specific topic either

... but either works :-)

**Comment by TinCanTech** [comment]
Sorry, I was not correcting you .. but I did see what _I thought_ you meant ;-)
